File: memOShield/memoshield/broadcaster.py
import threading
import queue
import json

class Broadcaster:
    def __init__(self):
        self._queues = []
        self._lock = threading.Lock()

    def register(self):
        q = queue.Queue()
        with self._lock:
            self._queues.append(q)
        return q

    def unregister(self, q):
        with self._lock:
            if q in self._queues:
                try:
                    self._queues.remove(q)
                except ValueError:
                    pass

    def publish(self, obj):
        data = json.dumps(obj, default=str)
        with self._lock:
            for q in list(self._queues):
                try:
                    q.put(data, block=False)
                except Exception:
                    # drop if queue full / closed
                    pass

# singleton
broadcaster = Broadcaster()

File: memOShield/internal/broadcaster/broadcaster.go
package broadcaster

import (
	"encoding/json"
	"sync"
)

// Broadcaster fans out events to registered SSE subscribers.
type Broadcaster struct {
	mu      sync.RWMutex
	clients []chan string
}

// Global singleton
var Global = &Broadcaster{}

// Register creates a new channel for a subscriber.
func (b *Broadcaster) Register() chan string {
	ch := make(chan string, 64)
	b.mu.Lock()
	b.clients = append(b.clients, ch)
	b.mu.Unlock()
	return ch
}

// Unregister removes and closes a subscriber channel.
func (b *Broadcaster) Unregister(ch chan string) {
	b.mu.Lock()
	defer b.mu.Unlock()
	for i, c := range b.clients {
		if c == ch {
			b.clients = append(b.clients[:i], b.clients[i+1:]...)
			close(ch)
			return
		}
	}
}

// Publish sends an event to all subscribers (non-blocking).
func (b *Broadcaster) Publish(obj interface{}) {
	data, err := json.Marshal(obj)
	if err != nil {
		return
	}
	s := string(data)
	b.mu.RLock()
	defer b.mu.RUnlock()
	for _, ch := range b.clients {
		select {
		case ch <- s:
		default:
			// drop if full
		}
	}
}
